httpclient5/src/main/java/org/apache/hc/client5/http/impl/routing/FailoverProxySelector.java

```diff
@@ -0,0 +1,127 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package org.apache.hc.client5.http.impl.routing;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.SocketAddress;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A ProxySelector implementation that selects proxies using a failover strategy.
+ * <p>
+ * This implementation maintains a list of other ProxySelectors, and selects a proxy
+ * by delegating to the next available ProxySelector in the list, until a non-empty
+ * proxy list is returned. If a ProxySelector fails with an exception, the next
+ * ProxySelector in the list is tried. Once a non-empty proxy list is returned by
+ * a ProxySelector, it becomes the active ProxySelector and is used for subsequent
+ * calls to select(). If a connection fails, connectFailed() is called on the
+ * active ProxySelector.
+ * <p>
+ * This implementation is useful in environments where multiple proxy servers may be
+ * used, and where it is desirable to failover to a different proxy if the current
+ * one fails.
+ *
+ * @since 5.3
+ */
+public class FailoverProxySelector extends ProxySelector {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FailoverProxySelector.class);
+
+    private final List<ProxySelector> selectors;
+    private final AtomicInteger currentIndex;
+
+
+    /**
+     * Constructs a FailoverProxySelector from the given list of ProxySelectors.
+     *
+     * @param selectors the list of ProxySelectors to use.
+     * @throws IllegalArgumentException if the list is null or empty.
+     */
+    public FailoverProxySelector(final List<ProxySelector> selectors) {
+        if (selectors == null || selectors.isEmpty()) {
+            throw new IllegalArgumentException("At least one ProxySelector is required");
+        }
+        this.selectors = new ArrayList<>(selectors);
+        currentIndex = new AtomicInteger(0);
+
+    }
+
+    /**
+     * Selects a list of proxies for the given URI by delegating to the next
+     * available ProxySelector in the list, until a non-empty proxy list is
+     * returned. Once a non-empty proxy list is returned, it becomes the active
+     * ProxySelector and is used for subsequent calls to select().
+     *
+     * @param uri the URI to select a proxy for.
+     * @return a list of proxies for the given URI.
+     */
+    @Override
+    public List<Proxy> select(final URI uri) {
```

Comment:
@arturobernalg What is exactly the intent of this implementation? Pick a backup selector in case the primary selector fails to produce a result or distribute the load randomly across multiple selectors? I am not quite sure I understand the logic of this method

Reply:
Is to select a list of proxies for the given URI by delegating to the next available ProxySelector in the list, until a non-empty proxy list is returned. It does not necessarily pick a backup selector in case the primary selector fails to produce a result, but rather rotates through the list of selectors until a non-empty proxy list is obtained

Reply:
@arturobernalg In this case `FailoverProxySelector` may be not the best choice of the name. I personally think `spraying` or  distributing` would likely be a better term.